### PR TITLE
Gradle plugin cleanup

### PR DESCRIPTION
**Pull request by shamsasari**
The Gradle plugin code has become complex and is hard to navigate. There's also code that is unnecessary, or not the "correct" way of doing things in Gradle. This clean up begins that process by making the code more manageable and easier to read. Further cleanup will probably be needed as `GradleEnclavePlugin` is still too big.

One of the main changes that has been done is making more use of Gradle task properties and providers and taking advantage of the auto-wiring of task dependencies. More info about this can be found here: https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:task_input_output_side_effects. A good example of this is the `EnclaveClassName` task, which now exposes an `enclaveClassName(): Provider<String>` function that will automatically wire-up the "depends on" relationship to any other task that needs the enclave class name.

There was also tasks that were either not used, or could be removed entirely.

### CODE REVIEW

**Comment by shamsasari** [comment]
@bonifazi-r3 or @ac101m can you you please review. I've marked this as draft to make sure it doesn't land before #81 .

@bonifazi-r3 what's the status of the QA tests? Do you want me to wait until they are green until we merge this?

File: plugin-enclave-gradle/src/main/kotlin/com/r3/conclave/plugin/enclave/gradle/EnclaveBundleJar.kt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        description = "Create Conclave bundle for $type mode"
+        archiveAppendix.set("bundle")
+        archiveClassifier.set(type.name.lowercase())
+    }
+
+    @get:InputFile
+    val bundleFile: RegularFileProperty = objects.fileProperty()
+
+    @get:Input
+    val fileName: Property<String> = objects.property(String::class.java)
```
Comment (bonifazi-r3) [inline plugin-enclave-gradle/src/main/kotlin/com/r3/conclave/plugin/enclave/gradle/EnclaveBundleJar.kt:25]: Isn't this an Output ? Just asking

File: plugin-enclave-gradle/src/main/kotlin/com/r3/conclave/plugin/enclave/gradle/EnclaveClassName.kt [hunk 2]
```diff
@@ -2,39 +2,48 @@ package com.r3.conclave.plugin.enclave.gradle
...
-    val inputClassPath: Property<FileCollection> = objects.property(FileCollection::class.java)
+    @get:InputFile
+    val sourceFatJar: RegularFileProperty = objects.fileProperty()
 
-    private val _outputEnclaveClassName = objects.property(String::class.java)
-    @get:Internal
-    val outputEnclaveClassName: Provider<String> get() = _outputEnclaveClassName
+    // Ideally we wouldn't write the output of the scan into a file, since doesn't need to be read outside of the
+    // plugin. However, Gradle tracks changes and up-to-date checks via files and directories, and so we write the
+    // result of the scan to a file which can then tracked by depdendent task. Notice how an @Output annotation
```
Comment (bonifazi-r3) [inline plugin-enclave-gradle/src/main/kotlin/com/r3/conclave/plugin/enclave/gradle/EnclaveClassName.kt:18]: Typo, `depdendent`

File: plugin-enclave-gradle/src/main/kotlin/com/r3/conclave/plugin/enclave/gradle/EnclaveFatJar.kt [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+        @get:Input
+        val enclaveClassName: Property<String> = objects.property(String::class.java)
+
+        @TaskAction
+        fun run() {
+            // Create the child copy spec before EnclaveFatJar executes
+            target.from(enclaveProperties.get()) { copySpec ->
+                copySpec.into(enclaveClassName.get().substringBeforeLast('.').replace('.', '/'))
+                copySpec.rename { PluginUtils.ENCLAVE_PROPERTIES }
```
Comment (bonifazi-r3) [inline plugin-enclave-gradle/src/main/kotlin/com/r3/conclave/plugin/enclave/gradle/EnclaveFatJar.kt:66]: Not an expert, but don't we miss an `Output` in this? I.e. my understanding is that when we create the task, at that moment we set the output (the name in this case).

File: plugin-enclave-gradle/src/main/kotlin/com/r3/conclave/plugin/enclave/gradle/GenerateEnclaveMetadata.kt [hunk 4]
```diff
@@ -34,8 +34,8 @@ open class GenerateEnclaveMetadata @Inject constructor(
         } else {
             commandLine(
                 plugin.signToolPath().absolutePathString(), "dump",
-                "-enclave", inputSignedEnclave.asFile.get(),
-                "-dumpfile", metadataFile
+                "-enclave", inputSignedEnclave.asFile.get().absolutePath,
+                "-dumpfile", metadataFile.absolutePathString()
```
Comment (bonifazi-r3) [inline plugin-enclave-gradle/src/main/kotlin/com/r3/conclave/plugin/enclave/gradle/GenerateEnclaveMetadata.kt:38]: Mmm, trying to understand... Don't we miss an @Output in this as well anyway?

File: plugin-enclave-gradle/src/main/kotlin/com/r3/conclave/plugin/enclave/gradle/GradleEnclavePlugin.kt [hunk 5]
```diff
@@ -115,53 +119,47 @@ class GradleEnclavePlugin @Inject constructor(private val layout: ProjectLayout)
...
+                    "Enclave revocation level not specified! " +
+                            "Please set the 'revocationLevel' property in the build configuration for your enclave.\n" +
+                            "If you're unsure what this error message means, please consult the conclave documentation.")
             }
         }
 
-        target.createTask<EnclaveClassName>("enclaveClassName") { task ->
-            task.dependsOn(target.tasks.withType(JavaCompile::class.java))
-            task.inputClassPath.set(getMainSourceSet(target).runtimeClasspath)
+        val sourceFatJar: Provider<RegularFile> = if (pythonSourcePath == null) {
```
Comment (bonifazi-r3) [inline plugin-enclave-gradle/src/main/kotlin/com/r3/conclave/plugin/enclave/gradle/GradleEnclavePlugin.kt:135]: I know that sounds silly, but why don't start to we put code like this this in functions? It is polluting

File: plugin-enclave-gradle/src/main/kotlin/com/r3/conclave/plugin/enclave/gradle/GradleEnclavePlugin.kt [hunk 6]
```diff
@@ -317,225 +232,206 @@ class GradleEnclavePlugin @Inject constructor(private val layout: ProjectLayout)
...
+        conclaveExtension: ConclaveExtension,
+        enclaveExtension: EnclaveExtension,
+        linuxExec: LinuxExec,
+        copyGraalVM: CopyGraalVM,
+        enclaveFatJarTask: EnclaveFatJar,
+        generateReflectionConfigTask: GenerateReflectionConfig,
+        generateAppResourcesConfigTask: GenerateAppResourcesConfig,
+        createDummyKeyTask: GenerateDummyMrsignerKey
+    ): Provider<RegularFile> {
+        fun buildFile(path: String): Provider<RegularFile> = file("${type.name.lowercase()}/$path")
```
Comment (bonifazi-r3) [inline plugin-enclave-gradle/src/main/kotlin/com/r3/conclave/plugin/enclave/gradle/GradleEnclavePlugin.kt:297]: I don't understand this provider mechanism here. Why don't we just set the outputs with a File?

**Review by bonifazi-r3 — CHANGES_REQUESTED** [review]
First round, few questions

File: plugin-enclave-gradle/src/main/kotlin/com/r3/conclave/plugin/enclave/gradle/GradleEnclavePlugin.kt [hunk 7]
```diff
@@ -317,225 +232,206 @@ class GradleEnclavePlugin @Inject constructor(private val layout: ProjectLayout)
...
-            val buildSignedEnclaveTask = target.createTask<BuildSignedEnclave>("buildSignedEnclave$type") { task ->
-                task.dependsOn(generateEnclaveMetadataTask)
-                task.inputs.files(generateEnclaveMetadataTask.inputSignedEnclave)
-                task.outputSignedEnclave.set(generateEnclaveMetadataTask.inputSignedEnclave)
+        val signedEnclaveFile: Provider<RegularFile> = enclaveExtension.signingType.flatMap {
+            when (it) {
+                SigningType.DummyKey -> signEnclaveWithKeyTask.outputSignedEnclave
+                SigningType.PrivateKey -> signEnclaveWithKeyTask.outputSignedEnclave
+                SigningType.ExternalKey -> addEnclaveSignatureTask.outputSignedEnclave
+                null -> throw IllegalStateException()  // Keep the compiler happy
```
Comment (ac101m) [inline plugin-enclave-gradle/src/main/kotlin/com/r3/conclave/plugin/enclave/gradle/GradleEnclavePlugin.kt:391]: Do we need this? This `when()` statement seems to be exhaustive

File: plugin-enclave-gradle/src/main/kotlin/com/r3/conclave/plugin/enclave/gradle/SignEnclave.kt [hunk 8]
```diff
@@ -28,9 +27,6 @@ open class SignEnclave @Inject constructor(
     @get:OutputFile
     val outputSignedEnclave: RegularFileProperty = objects.fileProperty()
```
Comment (ac101m) [inline plugin-enclave-gradle/src/main/kotlin/com/r3/conclave/plugin/enclave/gradle/SignEnclave.kt:28]: Missed one!

(input/output in variable name)

**Review by ac101m — COMMENTED** [review]
A few minor things

**Pull request closed without merging**